services/telemetry/telemetry_test.go

```diff
@@ -17,17 +17,114 @@
 package telemetry
 
 import (
+	"context"
+	"encoding/hex"
+	"os"
 	"testing"
+	"time"
 
+	"github.com/golang/protobuf/proto"
+	"github.com/google/uuid"
+	pmmv1 "github.com/percona-platform/saas/gen/telemetry/events/pmm"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
-func TestMakePayload(t *testing.T) {
-	s := NewService(nil, "1.3.1")
-	expected := "ECAB81E4C47D456CA9EC20AEBF91AB44;OS;\nECAB81E4C47D456CA9EC20AEBF91AB44;PMM;1.3.1\n"
-	assert.Equal(t, expected, string(s.makePayload("ECAB81E4C47D456CA9EC20AEBF91AB44"))) // \n are important
+const devTelemetryHost = "callhome-staging.percona.com:443"
+
+//nolint:lll
+func TestMakeV1Payload(t *testing.T) {
+	s := NewService(nil, "")
+
+	type param struct {
+		os      string
+		uuid    string
+		version string
+	}
+
+	for expected, p := range map[string]param{
+		"ACAB81E4C47D456CA9EC20AEBF91AB44;OS;linux\nACAB81E4C47D456CA9EC20AEBF91AB44;PMM;1.3.1\n": {os: "linux", uuid: "ACAB81E4C47D456CA9EC20AEBF91AB44", version: "1.3.1"},
+		"BCAB81E4C47D456CA9EC20AEBF91AB44;OS;ovf\nBCAB81E4C47D456CA9EC20AEBF91AB44;PMM;2.0.2\n":   {os: "ovf", uuid: "BCAB81E4C47D456CA9EC20AEBF91AB44", version: "2.0.2"},
+		"CCAB81E4C47D456CA9EC20AEBF91AB44;OS;ami\nCCAB81E4C47D456CA9EC20AEBF91AB44;PMM;2.3.0\n":   {os: "ami", uuid: "CCAB81E4C47D456CA9EC20AEBF91AB44", version: "2.3.0"},
+	} {
+		s.os = p.os
+		s.pmmVersion = p.version
+		actual := s.makeV1Payload(p.uuid)
+		assert.Equal(t, expected, string(actual)) // \n are important
+	}
+}
+
+func TestMakeV2Payload(t *testing.T) {
+	s := NewService(nil, "2.3.0")
+	delay := 6 * time.Hour
+	s.start = time.Now().Add(-delay)
+	s.tDistributionMethod = pmmv1.DistributionMethod_DOCKER
+	u, err := generateUUID()
+	require.NoError(t, err)
+
+	r, err := s.makeV2Payload(u)
+	require.NoError(t, err)
+	require.Len(t, r.Events, 1)
+
+	ev := r.Events[0]
+	assert.True(t, isValidUUID(t, ev.Id))
+
+	var uEv pmmv1.ServerUptimeEvent
+	err = proto.Unmarshal(ev.Event.Binary, &uEv)
+	require.NoError(t, err)
+
+	assert.Equal(t, uEv.Version, "2.3.0")
+	assert.Equal(t, uEv.DistributionMethod, pmmv1.DistributionMethod_DOCKER)
+	assert.LessOrEqual(t, float64(uEv.UpDuration.Seconds), (delay + 2*time.Second).Seconds())
+	assert.GreaterOrEqual(t, float64(uEv.UpDuration.Seconds), delay.Seconds())
+	assert.Equal(t, u, hex.EncodeToString(uEv.Id))
 }
 
+func TestSendV2Request(t *testing.T) {
+	os.Setenv(envV2Host, devTelemetryHost) //nolint:errcheck
+
+	t.Run("Normal", func(t *testing.T) {
+		s := NewService(nil, "2.3.0")
+		u, err := generateUUID()
+		require.NoError(t, err)
+		payload, err := s.makeV2Payload(u)
+		require.NoError(t, err)
+
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+
+		err = s.sendV2Request(ctx, payload)
```

Comment:
Is it a good idea to send request to a real server in test? Won't it make our tests unstable?

Reply:
That condition was written in the task. It is integration tests, in fact, so probably it's ok to make real request.

Reply:
The same thing as with tests in pmm-agent agains real databases instead of mocks – we want to know when things break, even if it is "not our fault, our tests pass".